File: examples/companion-packages/zig-zon-app/src/variants.zig
//! Typed enums for the four variant flags so the CLI can `switch` on them.

const std = @import("std");

pub const PetCardStyle = enum {
    compact,
    detailed,
    playful,

    pub fn parse(value: []const u8) ?PetCardStyle {
        if (std.mem.eql(u8, value, "compact")) return .compact;
        if (std.mem.eql(u8, value, "detailed")) return .detailed;
        if (std.mem.eql(u8, value, "playful")) return .playful;
        return null;
    }
};

pub const MatchStrategy = enum {
    popularity,
    match_quiz,
    longest_stay,

    pub fn parse(value: []const u8) ?MatchStrategy {
        if (std.mem.eql(u8, value, "popularity")) return .popularity;
        if (std.mem.eql(u8, value, "match-quiz")) return .match_quiz;
        if (std.mem.eql(u8, value, "longest-stay")) return .longest_stay;
        return null;
    }

    pub fn label(self: MatchStrategy) []const u8 {
        return switch (self) {
            .popularity => "popularity",
            .match_quiz => "match-quiz",
            .longest_stay => "longest-stay",
        };
    }
};

pub const MatchDepth = enum {
    short,
    standard,
    thorough,

    pub fn parse(value: []const u8) ?MatchDepth {
        if (std.mem.eql(u8, value, "short")) return .short;
        if (std.mem.eql(u8, value, "standard")) return .standard;
        if (std.mem.eql(u8, value, "thorough")) return .thorough;
        return null;
    }

    pub fn label(self: MatchDepth) []const u8 {
        return switch (self) {
            .short => "short",
            .standard => "standard",
            .thorough => "thorough",
        };
    }
};

pub const DonateFocus = enum {
    all,
    shelters,
    rescue,

    pub fn parse(value: []const u8) ?DonateFocus {
        if (std.mem.eql(u8, value, "all")) return .all;
        if (std.mem.eql(u8, value, "shelters")) return .shelters;
        if (std.mem.eql(u8, value, "rescue")) return .rescue;
        return null;
    }

    pub fn asStr(self: DonateFocus) []const u8 {
        return switch (self) {
            .all => "all",
            .shelters => "shelters",
            .rescue => "rescue",
        };
    }
};

File: examples/companion-packages/rust-cargo-app/src/variants.rs
//! Typed enums for the four variant flags so the CLI can `match` on them.

use std::fmt;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum PetCardStyle {
    Compact,
    Detailed,
    Playful,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum MatchStrategy {
    Popularity,
    MatchQuiz,
    LongestStay,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum MatchDepth {
    Short,
    Standard,
    Thorough,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum DonateFocus {
    All,
    Shelters,
    Rescue,
}

/// Error returned when a TinyFlags variant string does not map to any
/// known enum value.
#[derive(Debug, Clone)]
pub struct ParseVariantError {
    pub flag: &'static str,
    pub value: String,
}

impl fmt::Display for ParseVariantError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "pawmatch: unknown {} variant {:?}", self.flag, self.value)
    }
}

impl std::error::Error for ParseVariantError {}

impl PetCardStyle {
    pub fn parse(value: &str) -> Result<Self, ParseVariantError> {
        match value {
            "compact" => Ok(Self::Compact),
            "detailed" => Ok(Self::Detailed),
            "playful" => Ok(Self::Playful),
            _ => Err(ParseVariantError {
                flag: "pet-card-style",
                value: value.to_owned(),
            }),
        }
    }
}

impl MatchStrategy {
    pub fn parse(value: &str) -> Result<Self, ParseVariantError> {
        match value {
            "popularity" => Ok(Self::Popularity),
            "match-quiz" => Ok(Self::MatchQuiz),
            "longest-stay" => Ok(Self::LongestStay),
            _ => Err(ParseVariantError {
                flag: "recommendation-strategy",
                value: value.to_owned(),
            }),
        }
    }
}

impl fmt::Display for MatchStrategy {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let s = match self {
            Self::Popularity => "popularity",
            Self::MatchQuiz => "match-quiz",
            Self::LongestStay => "longest-stay",
        };
        f.write_str(s)
    }
}

impl MatchDepth {
    pub fn parse(value: &str) -> Result<Self, ParseVariantError> {
        match value {
            "short" => Ok(Self::Short),
            "standard" => Ok(Self::Standard),
            "thorough" => Ok(Self::Thorough),
            _ => Err(ParseVariantError {
                flag: "match-quiz-depth",
                value: value.to_owned(),
            }),
        }
    }
}

impl fmt::Display for MatchDepth {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let s = match self {
            Self::Short => "short",
            Self::Standard => "standard",
            Self::Thorough => "thorough",
        };
        f.write_str(s)
    }
}

impl DonateFocus {
    pub fn parse(value: &str) -> Result<Self, ParseVariantError> {
        match value {
            "all" => Ok(Self::All),
            "shelters" => Ok(Self::Shelters),
            "rescue" => Ok(Self::Rescue),
            _ => Err(ParseVariantError {
                flag: "donate-focus-default",
                value: value.to_owned(),
            }),
        }
    }

    pub fn as_str(self) -> &'static str {
        match self {
            Self::All => "all",
            Self::Shelters => "shelters",
            Self::Rescue => "rescue",
        }
    }
}
